shared/io/ktor/plugins/registry/PluginReference.kt

```diff
@@ -0,0 +1,226 @@
+    fun contains(other: ArtifactVersion): Boolean
+
+    val resolvedString: String get() =
+        asNumber()?.toString() ?: toString()
+    val safeName: String get() =
+        Regex("[^a-zA-Z0-9\\-,.]").replace(toString(), "").trim(',')
+}
+
+/**
+ * Special version string that ensures a plugin is the same as the ktor version.
+ */
+data class MatchKtor(val resolved: VersionNumber? = null) : ArtifactVersion {
+    override fun contains(other: ArtifactVersion) = true
+    override fun asNumber(): VersionNumber? = resolved
+    override fun resolve(version: VersionNumber): ArtifactVersion = MatchKtor(version)
+    override fun toString() = "=="
+}
+
+/**
+ * Standard semantic version number references (i.e., 1.0.0)
+ */
+data class VersionNumber(
+    val number: String,
+    val mavenVersion: org.apache.maven.artifact.versioning.DefaultArtifactVersion = org.apache.maven.artifact.versioning.DefaultArtifactVersion(
+        number
+    )
+) : ArtifactVersion, org.apache.maven.artifact.versioning.ArtifactVersion by mavenVersion {
+    override fun contains(other: ArtifactVersion): Boolean = this == other
+    override fun asNumber(): VersionNumber = this
+    override fun resolve(version: VersionNumber): ArtifactVersion = version
+    override fun toString(): String = number
+}
+
+data class VersionRange(
+    private val range: org.apache.maven.artifact.versioning.VersionRange,
+    private val resolved: VersionNumber? = null,
+) : ArtifactVersion {
+    constructor(text: String): this(org.apache.maven.artifact.versioning.VersionRange.createFromVersionSpec(text))
+    override fun contains(other: ArtifactVersion): Boolean = other is VersionNumber && range.containsVersion(other.mavenVersion)
+    override fun resolve(version: VersionNumber): ArtifactVersion = copy(resolved = version)
+    override fun asRange(): VersionRange = this
+    override fun asNumber(): VersionNumber? = resolved
+    override fun toString(): String = range.toString()
+}
+
+/**
+ * Variable name for version represented in catalog (i.e., $ktor_version)
+ */
+data class VersionVariable(val name: String, val version: ArtifactVersion): ArtifactVersion by version {
+    override fun resolve(version: VersionNumber): ArtifactVersion =
+        VersionVariable(name, version)
+
+    override fun toString() = "\$$name@$version"
+}
+
+/**
+ * This will replace prefix versions ending with .+ with the corresponding Maven version range.
+ * For example, 2.+ -> [2,3)
+ */
+fun prefixVersionToMavenRange(text: String): String {
+    require(text.endsWith(".+")) { "Expected \"$text\" to end with .+" }
+    val prefix = text.substringBeforeLast(".+")
+    val nextPrefix = prefix.split('.').mapIndexed { index, part ->
+        if (index < prefix.count { it == '.' }) part else (part.toInt() + 1).toString()
+    }.joinToString(".")
+    return "[$prefix,$nextPrefix)"
+}
+
+class ArtifactReferenceSerializer: kotlinx.serialization.KSerializer<ArtifactReference> {
```

Comment:
could you tell me why do we need package name here?

Reply:
I'll drop them here, Intellij added them as I was moving stuff around.